openprescribing/frontend/views/spending_utils.py

```diff
@@ -14,228 +20,351 @@
 NATIONAL_AVERAGE_DISCOUNT_PERCENTAGE = 7.2
 
 
+ConcessionPriceMatrices = namedtuple(
+    'ConcessionPriceMatrices',
+    'bnf_code_offsets date_offsets tariff_prices price_increases'
+)
+
+
+ConcessionCostMatrices = namedtuple(
+    'ConcessionCostMatrices',
+    'bnf_code_offsets date_offsets quantities tariff_costs extra_costs'
+)
+
+
 def ncso_spending_for_entity(entity, entity_type, num_months, current_month=None):
-    if entity_type.lower() == 'ccg':
-        prescribing_table = 'vw__presentation_summary_by_ccg'
-        entity_field = 'pct_id'
-        entity_condition = entity.code
-    elif entity_type == 'practice':
-        prescribing_table = 'frontend_prescription'
-        entity_field = 'practice_id'
-        entity_condition = entity.code
-    elif entity_type == 'all_england':
-        prescribing_table = 'vw__presentation_summary_by_ccg'
-        entity_field = 'frontend_pct.org_type'
-        entity_condition = 'CCG'
-    else:
-        raise ValueError('Unknown entity_type: '+entity_type)
+    org_type, org_id = _get_org_type_and_id(entity, entity_type)
     end_date = NCSOConcession.objects.aggregate(Max('date'))['date__max']
     # In practice, we always have at least one NCSOConcession object but we
     # need to handle the empty case in testing
     if not end_date:
         return []
-    start_date = end_date + relativedelta(months=-num_months)
-    with connection.cursor() as cursor:
-        sql, params = _ncso_spending_query(prescribing_table, start_date, end_date)
-        params.update(
-            entity_condition=entity_condition
-        )
-        cursor.execute("""
-            SELECT
-              month,
-              SUM(tariff_cost) AS tariff_cost,
-              SUM(additional_cost) AS additional_cost,
-              is_estimate,
-              %(last_prescribing_date)s AS last_prescribing_date
-            FROM
-              ({sql}) AS subquery
-            JOIN
-              frontend_pct ON frontend_pct.code = pct_id
-            WHERE
-              month > %(start_date)s AND month <= %(end_date)s
-              AND {entity_field} = %(entity_condition)s
-            GROUP BY
-              month, is_estimate
-            ORDER BY
-              month
-            """.format(
-                sql=sql, entity_field=entity_field),
-            params)
-        results = dictfetchall(cursor)
-    # Price concessions are released gradually over the course of the month.
-    # This means that the numbers for the current month will start off looking
-    # low and then gradually increase as more concessions are granted. We want
-    # to flag this to the user.
-    if current_month is not None:
-        for row in results:
-            row['is_incomplete_month'] = row['month'] >= current_month
+    start_date = end_date - relativedelta(months=num_months - 1)
+    last_prescribing_date = parse_date(get_db().dates[-1]).date()
```

Comment:
As I see it, the reliance on `ImportLog` was just a crutch because querying the actual source of truth for that fact was too slow under the old regime. But given that's not the case with the MatrixStore there's no reason to use it.

It would have the consequence that there were two independent bits of state to coordinate:  the version of the MatrixStore file being used and the contents of `ImportLog`. So you wouldn't e.g. just be able to grab and MatrixStore file and run against it locally without some sort of ceremony to get the `ImportLog` set up right.

Reply:
That makes sense but I still don't especiallylike having two inconsistent styles or at least not observable near each other in code - would it be nicer to have the two methods next to each other in utils or something (sorry I realise that would also be awkward, just writing in haste)